lib/ddtrace/workers/async.rb

```diff
@@ -0,0 +1,163 @@
+require 'ddtrace/logger'
+
+module Datadog
+  module Workers
+    module Async
+      # Adds threading behavior to workers
+      # to run tasks asynchronously.
+      # rubocop:disable Metrics/ModuleLength
+      module Thread
+        FORK_POLICY_STOP = :stop
+        FORK_POLICY_RESTART = :restart
+        SHUTDOWN_TIMEOUT = 1
+
+        def self.included(base)
+          base.send(:prepend, PrependedMethods)
+        end
+
+        # Methods that must be prepended
+        module PrependedMethods
+          def perform(*args)
+            start { self.result = super(*args) } if unstarted?
+          end
+        end
+
+        attr_reader \
+          :error,
+          :result
+
+        attr_writer \
+          :fork_policy
+
+        def join(timeout = nil)
+          return true unless running?
+          !worker.join(timeout).nil?
+        end
+
+        def terminate
+          return false unless running?
+          @run_async = false
+          worker.terminate
+          true
+        end
+
+        def run_async?
+          @run_async = false unless instance_variable_defined?(:@run_async)
+          @run_async == true
+        end
+
+        def unstarted?
+          worker.nil? || forked?
+        end
+
+        def running?
+          !worker.nil? && worker.alive?
+        end
+
+        def error?
+          @error = nil unless instance_variable_defined?(:@error)
+          !@error.nil?
+        end
+
+        def completed?
+          !worker.nil? && worker.status == false && !error?
+        end
+
+        def failed?
+          !worker.nil? && worker.status == false && error?
```

Comment:
I think the main purpose of this is just to indicate whether there's an `error` that can be read. Perhaps it should be expanded to the more general case though (status `nil`.)

Because we're catching `StandardError` it should never return status `nil` unless an exception is raised (e.g. thread abort.) Nonetheless maybe it should be `true` if aborted too.

I lean towards adopting this change.

Reply:
Made this change.